### PR TITLE
fix type instability of `Rprop`

### PR DESCRIPTION
**Pull request by MilkshakeForReal**
Now the type of states is always the same as the of parameters

### CODE REVIEW

File: src/rules.jl [hunk 1]
```diff
@@ -168,7 +168,8 @@ Rprop(η = 1f-3, ℓ = (5f-1, 1.2f0), Γ = (1f-6, 50f0)) = Rprop{typeof(η)}(η,
 init(o::Rprop, x::AbstractArray) = (zero(x), onevalue(o.eta, x))
 
 function apply!(o::Rprop, state, x, dx)
-    ℓ, Γ = o.ell, o.gamma
+    ℓ = o.ell .|> eltype(x)
+    Γ = o.gamma .|> eltype(x)
     g, η = state
```
Comment (mcabbott) [inline src/rules.jl:173]: This is the meat, `o.gamma` is a tuple. Would prefer `Γ = map(eltype(x), o.gamma)` but just style.

So the concern is that `Rprop(0.1)` will cause `(g, η)` to become Float64 arrays, I think. Can we test for this, ideally all of them? 

There are some tests here https://github.com/FluxML/Optimisers.jl/blob/master/test/runtests.jl#L120-L132 but they check only that the model's type isn't changed.

Comment (MilkshakeForReal) [reply src/rules.jl:173]: Should the returned `dx′` of `apply!` have the same element type with the input `x` or `dx`?

Comment (MilkshakeForReal) [reply src/rules.jl]: I have added some tests that should satisfy your requirement

File: test/runtests.jl [hunk 2]
```diff
@@ -129,8 +129,18 @@ Optimisers.trainable(x::TwoThirds) = (a = x.a,)
...
+
+      dm = ([1.0,2.0], [3.0,4.0])
+      s1 = Optimisers.setup(Rprop(0.1), m)
+      s2, m2 = Optimisers.update(s1, m, dm)
+      @test eltype(m2[1]) == Float16   # type should not be promoted by the gradient
+      @test eltype(m2[2]) == Float32
+      @test eltype(s2[1].state[1]) == eltype(s1[1].state[1])
+      @test eltype(s2[1].state[2]) == eltype(s1[1].state[2])
+      @test eltype(s2[2].state[1]) == eltype(s1[2].state[1])
+      @test eltype(s2[2].state[2]) == eltype(s1[2].state[2])
```
Comment (mcabbott) [inline test/runtests.jl:141]:
```suggestion
      # Rprop re-creates its state arrays, check they don't get widened:
      s5 = Optimisers.setup(Rprop(0.1), m)  # Float64 rule
      grad64 = ([1.0,2.0], SA[3.0,4.0])  # Float64 gradients
      s6, m6 = Optimisers.update(s5, m, grad64)
      @test eltype(m6[1]) == Float16
      @test eltype(m6[2]) == Float32
      @test eltype(s6[1].state[1]) == Float16
      @test eltype(s6[1].state[2]) == Float16
      @test eltype(s6[2].state[1]) == Float32
      @test eltype(s6[2].state[2]) == Float32
```

**Review by mcabbott — APPROVED** [review]
LGTM.

**Pull request merged**